design/service-apis.md

```diff
@@ -0,0 +1,249 @@
+# Service APIs
+
+This document outlines a design for implementing [Service APIs][1] in Contour Operator.
+
+## Goals
+
+- Manage Contour and dependent resources using Service APIs.
+
+## Non Goals
+
+- Support application-level Service APIs, i.e. xRoute. These APIs will be supported by Contour directly according to
+  Issue [2809][2].
+
+## Definitions
+
+### Service APIs
+
+Service APIs is an open source project managed by the Kubernetes SIG-NETWORK community. The project's goal is to evolve
+service networking APIs within the Kubernetes ecosystem. Service APIs provide interfaces to expose Kubernetes
+applications- Services, Ingress, and more.
+
+## Background
+
+Contour Operator manages Contour using the `contours.operator.projectcontour.io` Custom Resource Definition (CRD). The
+Contour Custom Resource (CR) provides the user interface for managing an instance of Contour and is handled by the
+Kubernetes API just like built-in objects. Contour Operator monitors the Contour CR and provides status on whether the
+actual state matches the desired state of the object. Contour Operator will continue to support the Contour CRD. Adding
+support for Service APIs will provide users an alternative approach to managing Contour.
+
+## High-Level Design
+
+Contour Operator will manage Contour using the `gatewayclasses.networking.x-k8s.io` and `gateways.networking.x-k8s.io`
+CRDs. The combination of GatewayClass and Gateway CRs will provide the user interface for managing an instance of
+Contour. Service APIs provides extension points to allow for custom resources to be linked at various layers of the API.
+This provides more granular customization at the appropriate places within the API structure. These extension points
+will be leveraged to expose Contour-specific configuration.
+
+## Detailed Design
+
+### GatewayClass
+
+GatewayClass is a cluster-scoped resource that defines a set of Gateways sharing a common behavior. A GatewayClass
+with `controller: projectcontour.io/contour-operator` will be managed by Contour Operator. There MUST be at least one
```

Comment:
nit: Should this just be `projectcontour.io/contour`? The Gateway is handled Contour not the operator.

Reply:
I guess the important bit is that is starts with `projectcontour.io`, the name after will be specific to the instance that should be managed, but all other names (i.e. not projectcontour.io will be ignored).

Reply:
> Should this just be projectcontour.io/contour? The Gateway is handled Contour not the operator.

This field represents the controller that manages the `Gateway`. This controller will reside in Contour Operator so IMO the value should indicate `contour-operator`. According to https://github.com/projectcontour/contour/pull/3213/files and https://github.com/projectcontour/contour/issues/2809, Contour will manage `xRoute` resources and not `Gateway` resources.

Reply:
I concur with @stevesloka. I think the class name should be `projectcontour.io/contour` to request that Contour satisfy the GatewayClass request. The fact that the operator provisions the class isn't relevant to the API.

Another way to say the same thing is that GatewayClass specifies the underlying infrastructure that gateways are scheduled onto. In this case, that infrastructure is Contour.

Reply:
@jpeach my understanding of the [spec](https://github.com/kubernetes-sigs/service-apis/blob/master/apis/v1alpha1/gatewayclass_types.go#L51-L52) is- who is responsible for managing Gateways of this class? The operator will be responsible for managing Gateways. Here's a sample use case:

1. A user creates GatewayClass/Gateway "Foo". Contour Operator creates the Contour environment and everything is great for a long time.
2. One day a new project called Foo comes along that is great at managing Contour using Service APIs.
3. The user is interested in creating a new Contour environment using project Foo. Since `projectcontour.io/contour` is used by Contour Operator and project Foo, they fight over which one is responsible for instantiating a Contour environment.

How can users differentiate using Contour Operator or another controller to manage Contour if both use the same `projectcontour.io/contour` value?

Reply:
It feels to like the `GatewayClass` should tell us this information (e.g. the operator is managing). The operator will react to any `Gateways` being created and deploy the Contour infrastructure, but at that point it's hands-off. Contur the controller then will watch for the specific Gateway it's responsible to manage (e.g. IngressClass) and process related `xRoute` resources. 

I think we can have quick discussion on this next time we're all together and hammer out an answer.